File: haskell/glue/src/Glue/Lib/IO/Print.hs
module Glue.Lib.IO.Print where

import Data.Text (unpack)
import Glue.Eval (Eval, liftIO)
import Glue.IR (IR (..))

printFunc :: IR Eval
printFunc = NativeFunc printFuncImpl

printFuncImpl :: IR Eval -> Eval (IR Eval)
printFuncImpl (String s) = liftIO (putStr (unpack s)) >> pure Void
printFuncImpl _ = pure Void

println :: IR Eval
println = NativeFunc printlnImpl

printlnImpl :: IR Eval -> Eval (IR Eval)
printlnImpl (String s) = liftIO (putStrLn (unpack s)) >> pure Void
printlnImpl _ = pure Void

File: dart/glue/lib/src/lib/io/print.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/ir.dart';

/// Print function - prints string without newline
/// Mirrors Haskell Glue.Lib.IO.Print.printFunc exactly
final Ir printFunc = IrNativeFunc(printFuncImpl);

/// Print function implementation
/// Mirrors Haskell Glue.Lib.IO.Print.printFuncImpl exactly
Eval<Ir> printFuncImpl(Ir arg) {
  print(arg);
  return Eval.pure(IrVoid());
}

/// Println function - prints string with newline
/// Mirrors Haskell Glue.Lib.IO.Print.println exactly
final Ir println = IrNativeFunc(printlnImpl);

/// Println function implementation
/// Mirrors Haskell Glue.Lib.IO.Print.printlnImpl exactly
Eval<Ir> printlnImpl(Ir arg) {
  print(arg);
  return Eval.pure(IrVoid()); // Haskell version ignores wrong arguments
}
